File: 32_Batch2_LegislativeAssembly/python/GoaLegislativeAssembly.py
# AUTHOR: PIYA JALMI
# ROLL NO: 32
# TITLE: GOA LEGISLATIVE ASSEMBLY
# START DATE: 8 JULY 2024
# MODIFIED DATE: 8 JULY 2024
# DESCRIPTION: THIS IS A MENU DRIVEN CODE OF GOA LEGISLATIVE ASSEMBLY WHICH HELPS YOU TO VIEW MLAS, ADD MLAS, VIEW WHAT ARE THE BILL INTRODUCED AND ALSO ISSUE A BILL

class Member:
    def __init__(self, name, contact_info):
        self.name = name
        self.contact_info = contact_info

    def get_name(self):
        return self.name

    def get_contact_info(self):
        return self.contact_info


class MLA(Member):
    def __init__(self, name, contact_info, constituency, party):
        super().__init__(name, contact_info)
        self.constituency = constituency
        self.party = party

    def get_constituency(self):
        return self.constituency

    def get_party(self):
        return self.party

    def introduce_bill(self, title, description, session):
        bill = Bill(title, description, self)
        self.constituency.add_bill(bill)
        session.add_bill(bill)
        print(f"Bill introduced successfully by {self.name}")


class Constituency:
    def __init__(self, name, area, population):
        self.name = name
        self.area = area
        self.population = population
        self.mlas = []
        self.bills = []

    def add_mla(self, mla):
        self.mlas.append(mla)

    def add_bill(self, bill):
        self.bills.append(bill)

    def get_name(self):
        return self.name

    def get_mlas(self):
        return self.mlas

    def get_bills(self):
        return self.bills


class PoliticalParty:
    def __init__(self, name, ideology, leader):
        self.name = name
        self.ideology = ideology
        self.leader = leader
        self.mlas = []

    def add_mla(self, mla):
        self.mlas.append(mla)

    def get_name(self):
        return self.name

    def get_mlas(self):
        return self.mlas


class Bill:
    def __init__(self, title, description, sponsor):
        self.title = title
        self.description = description
        self.sponsor = sponsor
        self.passed = False

    def get_title(self):
        return self.title

    def get_description(self):
        return self.description


class Session:
    def __init__(self, date):
        self.date = date
        self.bills = []

    def add_bill(self, bill):
        self.bills.append(bill)

    def get_date(self):
        return self.date

    def get_bills(self):
        return self.bills


def main():
    from datetime import date

    constituency1 = Constituency("Panaji", "Urban", 150000)
    constituency2 = Constituency("Margao", "Suburban", 200000)

    partyA = PoliticalParty("Party A", "Progressive", "Leader A")
    partyB = PoliticalParty("Party B", "Conservative", "Leader B")

    mla1 = MLA("MLA 1", "1234567890", constituency1, partyA)
    mla2 = MLA("MLA 2", "9876543210", constituency2, partyB)

    constituency1.add_mla(mla1)
    constituency2.add_mla(mla2)

    partyA.add_mla(mla1)
    partyB.add_mla(mla2)

    session1 = Session(date.today())
    session2 = Session(date.today())

    constituency1.add_bill(Bill("Education Reform Bill", "To improve educational standards", mla1))
    constituency2.add_bill(Bill("Healthcare Bill", "To enhance healthcare services", mla2))

    session1.add_bill(constituency1.get_bills()[0])
    session2.add_bill(constituency2.get_bills()[0])

    while True:
        print("\nGoa Legislative Assembly Menu:")
        print("1. View MLAs")
        print("2. View Bills in a Session")
        print("3. Introduce a Bill")
        print("4. Add MLAS")
        print("5. Exit")
        choice = int(input("Enter your choice: "))

        if choice == 1:
            print("\nMLAs in constituency 1:")
            for mla in constituency1.get_mlas():
                print(f"{mla.get_name()} - Constituency: {mla.get_constituency().get_name()}, Party: {mla.get_party().get_name()}")

            print("\nMLAs in constituency 2:")
            for mla in constituency2.get_mlas():
                print(f"{mla.get_name()} - Constituency: {mla.get_constituency().get_name()}, Party: {mla.get_party().get_name()}")

        elif choice == 2:
            print("\nBills in Session 1:")
            for bill in session1.get_bills():
                print(f"{bill.get_title()} - {bill.get_description()}")

        elif choice == 3:
            title = input("Enter bill title: ")
            description = input("Enter bill description: ")
            mla1.introduce_bill(title, description, session1)
            mla2.introduce_bill(title, description, session2)

        elif choice == 4:
            mla_name = input("Enter MLA name: ")
            contact_info = input("Enter contact info: ")
            constituency_choice = int(input("Enter constituency (1 for Panaji, 2 for Margao): "))
            party_choice = int(input("Enter party (1 for Party A, 2 for Party B): "))

            selected_constituency = constituency1 if constituency_choice == 1 else constituency2
            selected_party = partyA if party_choice == 1 else partyB

            new_mla = MLA(mla_name, contact_info, selected_constituency, selected_party)
            selected_constituency.add_mla(new_mla)
            selected_party.add_mla(new_mla)

            print("MLA added successfully.")

        elif choice == 5:
            print("Exiting...")
            break

        else:
            print("Invalid choice. Please enter again.")


if __name__ == "__main__":
    main()

File: 32_Batch2_LegislativeAssembly/src/GoaLegislativeAssembly.java
 //AUTHOR: PIYA JALMI
//ROLL NO: 32
// TITLE:GOA LEGISLATIVE ASSEMBLY
//START DATE:8 JULY 2024
//MODIFIED DATE:8 JULY 2024
//DESCRIPTION: THIS IS A MENU DRIVEN CODE OF GOA LEGISLATIVE ASSEMBLY WHICH HELPS YOU TO VIEW MLAS,ADD MLAS,VIEW WHAT ARE THE BILL INTRODUCED AND ALSO ISSUE A BILL

import java.util.*;

//THIS IS THE ABSTRACT CLASS WHICH PROVIDES GENERAL TEMPLATE FOR THE OTHER CLASSES
abstract class Member {
    protected String name;
    protected String contactInfo;

    public Member(String name, String contactInfo) {
        this.name = name;
        this.contactInfo = contactInfo;
    }
    public String getName() {
        return name;
    }
    public String getContactInfo() {
        return contactInfo;
    }
}
// THIS CLASS EXTENDS THE MEMBER CLASS INHERITING ITS ATTRIBUTES AND ALSO ADDS SOME ADDITIONAL ATTRIBUTES LIKE CONSTITUENCY AND POLITICAL PARTY WITH A METHOD INTORDUCE BILL
class MLA extends Member {
    private Constituency constituency;
    private PoliticalParty party;
    public MLA(String name, String contactInfo, Constituency constituency, PoliticalParty party) {
        super(name, contactInfo);
        this.constituency = constituency;
        this.party = party;
    }
    public Constituency getConstituency() {
        return constituency;
    }
    public PoliticalParty getParty() {
        return party;
    }
    public void introduceBill(String title, String description, Session session) {
        Bill bill = new Bill(title, description, this);
        constituency.addBill(bill);
        session.addBill(bill);
        System.out.println("Bill introduced successfully by " + name);
    }
}
//THIS CLASS REPRENSTS THE CONSTITUENCY WHICH WILL BE USED FURTHER WHILE ADDING AND VIEWING MLAS AND BILLS. IT USES ENCAPSULATION CONCEPT
class Constituency {
    private String name;
    private String area;
    private int population;
    private List<MLA> mlas;
    private List<Bill> bills;
    public Constituency(String name, String area, int population) {
        this.name = name;
        this.area = area;
        this.population = population;
        this.mlas = new ArrayList<>();
        this.bills = new ArrayList<>();
    }
    public void addMLA(MLA mla) {
        mlas.add(mla);
    }
    public void addBill(Bill bill) {
        bills.add(bill);
    }
    public String getName() {
        return name;
    }
    public List<MLA> getMLAs() {
        return mlas;
    }
    public List<Bill> getBills() {
        return bills;
    }
}
//THIS CLASS REPRENSTS THE POLITICALPARTY. IT USES ENCAPSULATION CONCEPT

class PoliticalParty {
    private String name;
    private String ideology;
    private String leader;
    private List<MLA> mlas;

    public PoliticalParty(String name, String ideology, String leader) {
        this.name = name;
        this.ideology = ideology;
        this.leader = leader;
        this.mlas = new ArrayList<>();
    }

    public void addMLA(MLA mla) {
        mlas.add(mla);
    }

    public String getName() {
        return name;
    }

    public List<MLA> getMLAs() {
        return mlas;
    }
}
//THIS CLASS REPRESENTS A BILL
class Bill {
    private String title;
    private String description;
    private MLA sponsor;
    private boolean passed;

    public Bill(String title, String description, MLA sponsor) {
        this.title = title;
        this.description = description;
        this.sponsor = sponsor;
        this.passed = false;
    }

   

    public String getTitle() {
        return title;
    }

    public String getDescription() {
        return description;
    }
}
//THIS CLASS REPRESENTS SESSION
class Session {
    private Date date;
    private List<Bill> bills;

    public Session(Date date) {
        this.date = date;
        this.bills = new ArrayList<>();
    }

    public void addBill(Bill bill) {
        bills.add(bill);
    }

    public Date getDate() {
        return date;
    }

    public List<Bill> getBills() {
        return bills;
    }
}
//THIS IS THE MAIN DRIVERE CLASS OF THE CODE WITH MENU DRIVER. USES INSTANCES OF OTHER CLASSES
public class GoaLegislativeAssembly {

    public static void main(String[] args) {
        Constituency constituency1 = new Constituency("Panaji", "Urban", 150000);
        Constituency constituency2 = new Constituency("Margao", "Suburban", 200000);

        PoliticalParty partyA = new PoliticalParty("Party A", "Progressive", "Leader A");
        PoliticalParty partyB = new PoliticalParty("Party B", "Conservative", "Leader B");

        MLA mla1 = new MLA("MLA 1", "1234567890", constituency1, partyA);
        MLA mla2 = new MLA("MLA 2", "9876543210", constituency2, partyB);

        constituency1.addMLA(mla1);
        constituency2.addMLA(mla2);

        partyA.addMLA(mla1);
        partyB.addMLA(mla2);

        Session session1 = new Session(new Date());
        Session session2 = new Session(new Date());

        constituency1.addBill(new Bill("Education Reform Bill", "To improve educational standards", mla1));
        constituency2.addBill(new Bill("Healthcare Bill", "To enhance healthcare services", mla2));

        session1.addBill(constituency1.getBills().get(0));
        session2.addBill(constituency2.getBills().get(0));

        Scanner scanner = new Scanner(System.in);
        int choice;
        do {
            System.out.println("\nGoa Legislative Assembly Menu:");
            System.out.println("1. View MLAs");
            System.out.println("2. View Bills in a Session");
            System.out.println("3. Introduce a Bill");
            System.out.println("4. Add MLAS");
            System.out.print("5.Exit ");
            System.out.print("6.Enter your choice: ");

            choice = scanner.nextInt();

            switch (choice) {
                case 1:
                    System.out.println("\nMLAs in constituency 1:");
                    for (MLA mla : constituency1.getMLAs()) {
                        System.out.println(mla.getName() + " - Constituency: " + mla.getConstituency().getName()
                                + ", Party: " + mla.getParty().getName());
                    }
                    System.out.println("\nMLAs: in constituency 2");
                    for (MLA mla : constituency2.getMLAs()) {
                        System.out.println(mla.getName() + " - Constituency: " + mla.getConstituency().getName()
                                + ", Party: " + mla.getParty().getName());
                    }
                    break;
                case 2:
                    System.out.println("\nBills in Session 1:");
                    for (Bill bill : session1.getBills()) {
                        System.out.println(bill.getTitle() + " - " + bill.getDescription());
                    }
                    break;
                    case 3:
                    scanner.nextLine(); 
                    System.out.print("Enter bill title: ");
                    String title = scanner.nextLine();
                    System.out.print("Enter bill description: ");
                    String description = scanner.nextLine();
                    mla1.introduceBill(title, description, session1); 
                    mla2.introduceBill(title, description, session2); 

                    break;
                    case 4:
                    scanner.nextLine(); 
                    System.out.print("Enter MLA name: ");
                    String mlaName = scanner.nextLine();
                    System.out.print("Enter contact info: ");
                    String contactInfo = scanner.nextLine();
                    System.out.print("Enter constituency (1 for Panaji, 2 for Margao): ");
                    int constituencyChoice = scanner.nextInt();
                    System.out.print("Enter party (1 for Party A, 2 for Party B): ");
                    int partyChoice = scanner.nextInt();

                    Constituency selectedConstituency = constituencyChoice == 1 ? constituency1 : constituency2;
                    PoliticalParty selectedParty = partyChoice == 1 ? partyA : partyB;

                    MLA newMLA = new MLA(mlaName, contactInfo, selectedConstituency, selectedParty);
                    selectedConstituency.addMLA(newMLA);
                    selectedParty.addMLA(newMLA);

                    System.out.println("MLA added successfully.");
                    break;
                case 5:
                    System.out.println("Exiting...");
                    break;
                default:
                    System.out.println("Invalid choice. Please enter again.");
                    break;
            }
        } while (choice != 5);

        scanner.close();
    }
}
